### PR TITLE
QSG-Shell scripts

### PR DESCRIPTION
**Pull request by NipuniBhagya**
This includes shell scripts to automate the following.
1. Configuring SAML2 web SSO
2. Configuring OIDC web SSO
3. Configuring Twitter as a Federated Authenticator

### CODE REVIEW

File: QSG-pack/src/main/resources/QSG/bin/02/response_unformatted.xml [hunk 1]
```diff
@@ -0,0 +1 @@
...thenticationConfig"><ax2140:alwaysSendBackAuthenticatedListOfIdPs>false</ax2140:alwaysSendBackAuthenticatedListOfIdPs><ax2140:authenticationStepForAttributes xsi:nil="true"/><ax2140:authenticationStepForSubject xsi:nil="true"/><ax2140:authenticationType>default</ax2140:authenticationType><ax2140:enableAuthorization>false</ax2140:enableAuthorization><ax2140:subjectClaimUri xsi:nil="true"/><ax2140:useTenantDomainInLocalSubjectIdentifier>false</ax2140:useTenantDomainInLocalSubjectIdentifier><ax2140:useUserstoreDomainInLocalSubjectIdentifier>false</ax2140:useUserstoreDomainInLocalSubjectIdentifier></ax2140:localAndOutBoundAuthenticationConfig><ax2140:outboundProvisioningConfig xsi:type="ax2140:OutboundProvisioningConfig"><ax2140:provisionByRoleList xsi:nil="true"/></ax2140:outboundProvisioningConfig><ax2140:owner xsi:type="ax2140:User"><ax2140:tenantDomain>carbon.super</ax2140:tenantDomain><ax2140:userName>cameron</ax2140:userName><ax2140:userStoreDomain>PRIMARY</ax2140:userStoreDomain></ax2140:owner><ax2140:permissionAndRoleConfig xsi:type="ax2140:PermissionsAndRoleConfig"/><ax2140:saasApp>false</ax2140:saasApp></ns:return></ns:getApplicationResponse></soapenv:Body></soapenv:Envelope>
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/resources/QSG/bin/02/response_unformatted.xml]: Add a new line to the file

File: QSG-pack/src/main/resources/QSG/bin/05/delete-idp-twitter.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+-->
+<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:mgt="http://mgt.idp.carbon.wso2.org">
+    <soapenv:Header/>
+    <soapenv:Body>
+        <mgt:deleteIdP>
+            <!--Optional:-->
+            <mgt:idPName>IDP-twitter</mgt:idPName>
+        </mgt:deleteIdP>
+    </soapenv:Body>
+</soapenv:Envelope>
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/resources/QSG/bin/05/delete-idp-twitter.xml]: Add a new line please

File: QSG-pack/src/main/resources/QSG/bin/05/response_unformatted.xml [hunk 3]
```diff
@@ -0,0 +1 @@
...ckAuthenticatedListOfIdPs>false</ax2140:alwaysSendBackAuthenticatedListOfIdPs><ax2140:authenticationStepForAttributes xsi:nil="true"/><ax2140:authenticationStepForSubject xsi:nil="true"/><ax2140:authenticationType>default</ax2140:authenticationType><ax2140:enableAuthorization>false</ax2140:enableAuthorization><ax2140:subjectClaimUri>http://wso2.org/claims/fullname</ax2140:subjectClaimUri><ax2140:useTenantDomainInLocalSubjectIdentifier>false</ax2140:useTenantDomainInLocalSubjectIdentifier><ax2140:useUserstoreDomainInLocalSubjectIdentifier>false</ax2140:useUserstoreDomainInLocalSubjectIdentifier></ax2140:localAndOutBoundAuthenticationConfig><ax2140:outboundProvisioningConfig xsi:type="ax2140:OutboundProvisioningConfig"><ax2140:provisionByRoleList xsi:nil="true"/></ax2140:outboundProvisioningConfig><ax2140:owner xsi:type="ax2140:User"><ax2140:tenantDomain>carbon.super</ax2140:tenantDomain><ax2140:userName>cameron</ax2140:userName><ax2140:userStoreDomain>PRIMARY</ax2140:userStoreDomain></ax2140:owner><ax2140:permissionAndRoleConfig xsi:type="ax2140:PermissionsAndRoleConfig"/><ax2140:saasApp>false</ax2140:saasApp></ns:return></ns:getApplicationResponse></soapenv:Body></soapenv:Envelope>
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/resources/QSG/bin/05/response_unformatted.xml]: Add a new  line

File: QSG-pack/src/main/resources/QSG/bin/qsg.sh [hunk 4]
```diff
@@ -0,0 +1,1053 @@
+#!/bin/sh
+
+# ----------------------------------------------------------------------------
+#  Copyright 2017 WSO2, Inc. http://www.wso2.org
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/resources/QSG/bin/qsg.sh]: Can we change this license header to 2018?

File: QSG-pack/src/main/java/AppId.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+public class AppId {
+
+    private static Logger LOGGER = Logger.getLogger(AppId.class.getSimpleName());
+
+    public static void main(String[] args) {
+
+        try {
+            DocumentBuilderFactory docBuilderFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
+            Document doc = docBuilder.parse(new File("response_unformatted.xml"));
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/java/AppId.java]: Shall we use a priave static final String RESPONSE_FORMATTED_FILE = "response_unformatted.xml"

File: QSG-pack/src/main/java/AppId.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+            NodeList appId = doc.getElementsByTagName("ax2140:applicationID");
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/java/AppId.java]: Put this tag also to a constant or  a final string as above

File: QSG-pack/src/main/java/AppId.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+            Element appElement = (Element) appId.item(0);
+            NodeList appList = appElement.getChildNodes();
+            System.out.println(((Node) appList.item(0)).getNodeValue().trim());
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/java/AppId.java]: Better remove all souts and put a debug logs or a info logs for them. This is not a best practice.

Comment (NipuniBhagya) [reply QSG-pack/src/main/java/AppId.java]: Actually, this is to be read by the script. So I think we'll have to keep this. :)

File: QSG-pack/pom.xml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    <groupId>Quick-Start-Guide</groupId>
+    <artifactId>QSG</artifactId>
+    <version>1.0</version>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+                <version>3.0.2</version>
```
Comment (GDRDABARERA) [inline QSG-pack/pom.xml]: Shall we move this versions to master pom.xml under <properties>, and use them here as ${maven.plugin.version}.
Do this for other versions as well

File: QSG-pack/src/main/java/AppId.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    public static void main(String[] args) {
+
+        try {
+            DocumentBuilderFactory docBuilderFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder docBuilder = docBuilderFactory.newDocumentBuilder();
+            Document doc = docBuilder.parse(new File("response_unformatted.xml"));
+
+            NodeList appId = doc.getElementsByTagName("ax2140:applicationID");
+
+            Element appElement = (Element) appId.item(0);
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/java/AppId.java]: Can appElement become null or emplty? If so check  null here, appElement.getChildNodes();

File: QSG-pack/src/main/resources/QSG/bin/qsg.sh [hunk 10]
```diff
@@ -0,0 +1,1053 @@
+#!/bin/sh
```
Comment (GDRDABARERA) [inline QSG-pack/src/main/resources/QSG/bin/qsg.sh]: What if a user ran the same thing twice, will he get errors like "dupliacate SP" ?, or Is there a place to revert the changes?

Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/qsg.sh]: The setup will be cleaned after every try. Therefore all the users, roles and SPs will be deleted at the end of each run.

Comment (GDRDABARERA) [reply QSG-pack/src/main/resources/QSG/bin/qsg.sh]: What will happen if corrupted in the middle?

File: QSG-pack/src/main/resources/QSG/bin/02/response_unformatted.xml (hunk 1 above)
Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/02/response_unformatted.xml]: This is file contains a SOAP response. Therefore we can't add new lines manually.

File: QSG-pack/src/main/resources/QSG/bin/05/response_unformatted.xml (hunk 3 above)
Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/05/response_unformatted.xml]: This is file also contains a SOAP response. Therefore we can't add new lines manually.

File: QSG-pack/src/main/resources/QSG/bin/05/delete-idp-twitter.xml (hunk 2 above)
Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/05/delete-idp-twitter.xml]: Added a new line.

File: QSG-pack/src/main/resources/QSG/bin/qsg.sh (hunk 10 above)
Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/qsg.sh]: The script will exit and return -1.

Comment (GDRDABARERA) [reply QSG-pack/src/main/resources/QSG/bin/qsg.sh]: Does the changes are revert back?

Comment (NipuniBhagya) [reply QSG-pack/src/main/resources/QSG/bin/qsg.sh]: The setup will be cleaned up to that point.

File: QSG-pack/src/main/java/AppId.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+                // This stdout is added to be read by the shell script.
+                System.out.println(((Node) appList.item(0)).getNodeValue().trim());
+
+            } else {
+                LOGGER.info("Application ID not found.");
+            }
+
+        } catch (SAXParseException err) {
+            LOGGER.info("** Parsing error" + ", line "
```
Comment (hasinthaindrajee) [inline QSG-pack/src/main/java/AppId.java]: Aren't these should be error logs with stacktraces ?. This way we can't really say what the cause is since we don't print it. Isn't it ?
